CDAP-4294 namespace isolation for system artifact plugins

albertshau opened the pull request:
Fix a bug where users in one namespace could see plugins from
another namespace when using a system application artifact.
To fix this, when deploying an app, we need to pass down the
namespace the app will be in to filter out plugin artifacts from
other namespaces. Most of the changes revolve around passing
namespace to the configurers that require it.

albertshau commented:
http://builds.cask.co/browse/CDAP-RBT586

gokulavasan commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/InMemoryConfigurator.java:
```diff
@@ -79,17 +80,19 @@
   private Id.Artifact artifactId;
 
-  public InMemoryConfigurator(CConfiguration cConf, Id.Artifact artifactId, String appClassName,
+  public InMemoryConfigurator(CConfiguration cConf, Id.Namespace namespace, Id.Artifact artifactId, String appClassName,
```
Hmm, wondering the field names can be better. For example, what is the difference between ``namespace`` and ``artifactId.getNamespaceId()`` ? Without context, ``namespace`` variable seems redundant.

gokulavasan commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java:
```diff
@@ -508,6 +512,12 @@ public ArtifactData apply(DatasetContext<Table> context) throws Exception {
             // column is the artifact name and version, value is the serialized PluginClass
             for (Map.Entry<byte[], byte[]> column : row.getColumns().entrySet()) {
               ArtifactColumn artifactColumn = ArtifactColumn.parse(column.getKey());
+              Id.Namespace artifactNamespace = artifactColumn.artifactId.getNamespace();
```
Not sure if we had discussed this before, but there seems to be quite a bit of code duplication. Anyway to avoid it?

gokulavasan commented:
Couple of comments. Rest LGTM :+1:

gokulavasan commented:
Just a thought, wondering about having unit tests in three different places. Are they testing different features? Wondering if we are covering different parts of the codebase.

albertshau commented:
@gokulavasan addressed comments, please take another look.  The tests are covering slightly different things. For example, I was originally passing the wrong namespace in the http handlers, which wasn't caught by the ArtifactStore and ArtifactRepository tests.

gokulavasan commented:
LGTM :+1:

albertshau commented:
squashed commits
